test/tools/yulopti.cpp

```diff
@@ -171,28 +206,138 @@ class YulOpti
 		}
 	}
 
-	void disambiguate()
+	void applyFunctionToObject(shared_ptr<Object> _object, function<void(Object&)> _function)
 	{
-		*m_ast = std::get<yul::Block>(Disambiguator(m_dialect, *m_analysisInfo)(*m_ast));
-		m_analysisInfo.reset();
-		m_nameDispenser.reset(*m_ast);
+		for (auto const& subObjectNode: _object->subObjects) {
+			auto subObject = dynamic_pointer_cast<Object>(subObjectNode);
+
+			if (subObject != nullptr)
+				applyFunctionToObject(subObject, _function);
+		}
+
+		_function(*_object);
 	}
 
-	void runSteps(string _source, string _steps)
+	void runCodeAnalyzer(ErrorReporter& _errorReporter)
 	{
-		parse(_source);
-		disambiguate();
-		OptimiserSuite{m_context}.runSequence(_steps, *m_ast);
-		cout << AsmPrinter{m_dialect}(*m_ast) << endl;
+		applyFunctionToObject(
+			m_object,
+			[&](Object& _object)
+			{
+				_object.analysisInfo = make_shared<yul::AsmAnalysisInfo>();
+
+				AsmAnalyzer analyzer(
+					*_object.analysisInfo,
+					_errorReporter,
+					m_dialect,
+					{},
+					_object.qualifiedDataNames()
+				);
+
+				bool success = analyzer.analyze(*_object.code);
+				yulAssert(success && !_errorReporter.hasErrors(), "Invalid assembly/yul code.");
+			}
+		);
 	}
 
-	void runInteractive(string _source, bool _disambiguated = false)
+	void runCodeDisambiguator()
 	{
+		applyFunctionToObject(
+			m_object,
+			[&](Object& _object)
+			{
+				_object.code = make_shared<yul::Block>(
+					get<yul::Block>(Disambiguator(m_dialect, *_object.analysisInfo)(*_object.code))
+				);
+
+				_object.analysisInfo.reset();
+			}
+		);
+	}
+
+	void runSequence(string_view _steps)
+	{
+		applyFunctionToObject(
+			m_object,
+			[&](Object& _object)
+			{
+				OptimiserSuite{*m_context}.runSequence(_steps, *_object.code);
+			}
+		);
+	}
```

Comment:
I recently realized that there's a big problem here. This code ignores the distinction between *creation* and *deployed* objects (see [Specification of Yul Object](https://docs.soliditylang.org/en/latest/yul.html#specification-of-yul-object)). Normally the optimizer considers the top-level object and any objects with names not ending with `_deployed` to be creation objects. For these `expectedExecutionsPerDeployment` (i.e. `--optimize-runs`) is set to 1 and it affects the result of the optimization. This logic sits in `YulStack::optimize()` but the code here ignores it and just walks the object treating every sub-object the same way.

This wasn't a problem as long as we were working with only one object in isolation. With this PR it becomes misleading because when you pass it the whole object you'll get slightly different output than you would get from the compiler.

Reply:
I see. Would it make sense to call `YulStack::optimize()` here instead or should I implement the same logic?

I'm assuming that the top-level object will be the top-level object from the input and never the object given by `--object`?

Reply:
I'd rather avoid reimplementing all of it because we'll easily forget to update yulopti if the code diverges. On the other hand we already make assumptions about what the optimizer does so I don't think we can fully avoid that.

The part I'd like to avoid duplicating the most is the logic for determining if an object is a creation object. I think it would be acceptable to keep the current solution of walking the object on your own if you extract a common method with logic for determining it and make both yulopti and `YulStack` use it. You could put it in `Object`.

Also, I'd recommend doing this (and the refactor I suggested in https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654) in a separate PR. Changes outside of yulopti (which is just a non-user-facing testing tool) have more potential to break things so would be better to have them separately and then to get more people to look specifically at them.

Reply:
> I'm assuming that the top-level object will be the top-level object from the input and never the object given by `--object`?

We could do it either way - with the top being fixed or with it being the subobject one specified by `--object`- but I think the former would be more intuitive. I.e. we always optimize the whole object and `--object` works simply as a filter to show you just the one object you care about - as opposed to cutting out the subobject and optimizing it as if it was the whole thing (which you can always achieve just by modifying the file).

It would be a good idea to say that in the help string of the option.

Reply:
I've updated the PR, adding proper handling of creation objects. For now I've duplicated the check code. I'm not sure how the check can be properly refactored into `Object` since knowing whether an object is a creation object or not requires knowing if the object is at the root of the tree. Some possible method declarations would be:
```
bool Object::isCreation(string _qualifiedPath);
bool Object::isCreation(int depth);
bool Object::isCreation(bool isRootObject);
```

Maybe just `bool Object::isDeployed()` would do?

I had done the refactor suggested in [#12941 (comment)](https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654) on this PR branch previously. I can of course move it into a separate PR.

Reply:
> ```solidity
> bool Object::isCreation(bool isRootObject);
> ```

I'd go with this one. The one taking the path might actually be nicer but in some places where we need this (in particular `YulStack`), we don't have the path at hand so we'd have spend extra effort to construct it. We could also add both, one using the other under the hood, but that's optional.

> Maybe just `bool Object::isDeployed()` would do?

If you mean a method only checking for `_deployed` in the name, I think that would be a bit misleading because for the top-level object it could return `true` even if it's a creation object. I see `isDeployed` as the opposite of `isCreation`.

> I had done the refactor suggested in https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654 on this PR branch previously. I can of course move it into a separate PR.

I'd really recommend moving it to `Object` and doing that in a separate PR. We have several places that have this condition hard-coded and I think it would be best if they all used a single method. To be honest, I'm not even confident that they deal with the top-level object the same way so it would be a good opportunity to check that and maybe add some tests to cover it.

[Specification of Yul Object](https://docs.soliditylang.org/en/latest/yul.html#specification-of-yul-object) does not even mention the top-level object so we could also update it in that PR.